docs: Add Kubernetes section to the architecture document

sameo opened the pull request:
We describe the relationship between CRI-O and CC, and how they
work together to properly support VM based runtimes for Kubernetes.

Fixes #726

Signed-off-by: Samuel Ortiz <sameo@linux.intel.com>

sboeuf commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+code and thus from a security perspective it makes sense to tag them as `untrusted`. A `CRI-O`
+and Clear Containers based Kubernetes cluster handles this use case transparently as long as
+the deployed containers are properly tagged. All `untrusted` containers will be handled by
+Clear Container and thus run in a hardware virtualized secure sandbox while `runc` will handle
```
s/Container/Containers/

jodh-intel replied:
"whilst `runc`, for example, could handle ..." maybe to allow for other runtimes?

sboeuf commented:
Except one comment, this
LGTM

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
 of the underlying block device through the runtime configuration.
 
+## Kubernetes support
+[Kubernetes\*](https://github.com/kubernetes/kubernetes/) is the dominant open source
```
> Kubernetes is the dominant open source ...

This may be true, but I wouldn't mention that for 2 reasons:

- it may not be true tomorrow :)
- it's arguably an opinion.

grahamwhaley replied:
Agree. Maybe s/dominant/popular/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+[CRI-O\*](https://github.com/kubernetes-incubator/cri-o) is a `CRI` implementation
+that relies on `OCI` compatible runtimes for managing containers instances.
+
+Clear Containers being an OCI compatible runtime, it naturally fits into the `CRI-O`
```
"Clear Containers, being an OCI compatible runtime, naturally fits ..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+that relies on `OCI` compatible runtimes for managing containers instances.
+
+Clear Containers being an OCI compatible runtime, it naturally fits into the `CRI-O`
+architecture and requirements, and is one of the 2 officially supported `CRI-O` runtimes.
```
s/2/two/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+### Problem statement
+
+[Kubernetes](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
```
"The Kubernetes execution unit ..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+[Kubernetes](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
+i.e. a set of containers sharing the same resources and networking, mount and PID
+namespaces. A pod specification is a set of containers associated to a list of
```
"is a set of containers with an associated list of constraints".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+namespaces. A pod specification is a set of containers associated to a list of
+constraints (namespaces, cgroups, hardware resources, security contexts, etc...).
+By default the Clear Containers runtime associates each container with a dedicated
+virtual machine, but that does not work really well with pods as letting a set of
```
I'd remove the word "really".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+By default the Clear Containers runtime associates each container with a dedicated
+virtual machine, but that does not work really well with pods as letting a set of
+VMs share the same namespaces and resources can be very complex to implement. For
+example allowing 2 virtual machines to share the same networking or PID namespace
```
s/2/two/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+but with Kubernetes it will create and launch one virtual machine per pod. Then each
+container within a pod is simply a container inside the VM boundary.
+
+So the challenge with Clear Containers when trying to work as a `CRI-O` runtime is
```
"The challenge for Clear Containers when working as a ..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+So the challenge with Clear Containers when trying to work as a `CRI-O` runtime is
+to know when to create a full virtual machine (for pods) and when to create a new
+container inside a previously created VM. In both cases it will get called with very
+similar arguments, so it needs `CRI-O` help to be able to distinguish between a pod
```
"so it needs the help of `CRI-O` to be able to ..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+### OCI annotations
+
+In order for the Clear Containers runtime (Or any VM based OCI compatible runtime) to
```
s/Or/or/

grahamwhaley replied:
You can probably drop 'any VM based' - as the method I think is not tied to VM based runtimes, so could work for other non-VM use cases as well in theory.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+In order for the Clear Containers runtime (Or any VM based OCI compatible runtime) to
+be able to understand if it needs to create a full virtual machine or if it has to create
+a new container inside an existing pod's VM context, `CRI-O` adds specific annotations
+to the OCI configuration file (`config.json`) passed down to the OCI compatible runtimes
```
I'd just say "(`config.json`) which is passed to the OCI compatible runtime."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+it relies on.
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` OCI configuration files it produces from the kubelet CRI
```
s/configuration files/configuration file/

grahamwhaley replied:
We may need to backtick all ocurrences of `OCI` and `CRI` throughout the new section. Need clarification on if that is necessary - at present we have a mix of some backticked and some not.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` OCI configuration files it produces from the kubelet CRI
+request. `io.kubernetes.cri-o.ContainerType` can either be set to `sandbox` or `container`.
```
"The `io.kubernetes.cri-o.ContainerType` annotation can either ..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+### Generalization
+
+With Clear Containers 2.1, we implemented the initial Kubernetes/`CRI-O` support by choosing
```
- Missing spaces either side of the slash.
- I'd write this impersonally maybe ("Clear Containers 2.1 implemented the initial ...").

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+With Clear Containers 2.1, we implemented the initial Kubernetes/`CRI-O` support by choosing
+to handle pods as a special case for containers. By default the runtime would only be aware
+of containers and only when receiving CRI-O specific annotations it would follow specific
```
It reads better if you change:

- s/it would/would it/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+to handle pods as a special case for containers. By default the runtime would only be aware
+of containers and only when receiving CRI-O specific annotations it would follow specific
+code paths to handle pods and intra pod containers. This approach showed that container
+orchestration support was an after thought with 2.1, and that supporting the single container
```
s/after thought/afterthought/

Ooh an Oxford Comma ;)

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+orchestration support was an after thought with 2.1, and that supporting the single container
+Docker use case was the main focus.
+
+Starting with Clear Containers 3.0 we took a different, almost opposite approach. The Clear
```
Again, I'd be tempted to remove the "we".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+Docker use case was the main focus.
+
+Starting with Clear Containers 3.0 we took a different, almost opposite approach. The Clear
+Containers 3.0 runtime is based on the `virtcontainers` package which API is inspired by the
```
s/which API/whose API/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+Containers 3.0 runtime is based on the `virtcontainers` package which API is inspired by the
+Kubernetes `CRI` one. `virtcontainers` execution units are pods and it does not create or manage
+any container outside of an existing pod context, and thus so does the Clear Container 3.0 runtime.
+To create a Docker container it will first create a pod which will represent the container
```
To me it reads better if you reorder the words and refer to "the hardware virtualized container context".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+any container outside of an existing pod context, and thus so does the Clear Container 3.0 runtime.
+To create a Docker container it will first create a pod which will represent the container
+hardware virtualized context and then start an actual container inside that virtual machine.
+Both the pod and its container have the same id and name.
```
s/id/identifier/ maybe?

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+### Mixing VM based and namespace based runtimes
+
+One interesting evolution of the `CRI-O` support for the Clear Containers runtime is the ability
+to run VM based pods alongside with namespace ones. With `CRI-O` and Clear Containers, one can
```
I'd remove "with" as it's redundant here.

grahamwhaley commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+agent called `kubelet`. On each compute node a `kubelet` instance is responsible
+for managing the node pods lifecycle and eventually relies on a container runtime for
+handling the container execution parts. The `kubelet` architecture decouples the
+lifecycle management part from the container execution one through the dedicated `gRPC`
```
consider hot-linking `gRPC` to maybe the gRPC definition or docs (either generic google ones or specific k8s/CRI-O ones if such a thing exists)

grahamwhaley commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+annotation to the `config.json` OCI configuration files it produces from the kubelet CRI
+request. `io.kubernetes.cri-o.ContainerType` can either be set to `sandbox` or `container`.
+Clear Containers will then use this annotation to decide if it needs to respectively
+create a virtual machine or a container inside a virtual machine associated with a
```
we have a mix of 'virtual machine' and 'VM' through the section - we may need to make that consistent

grahamwhaley commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,117 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+and Clear Containers based Kubernetes cluster handles this use case transparently as long as
+the deployed containers are properly tagged. All `untrusted` containers will be handled by
+Clear Container and thus run in a hardware virtualized secure sandbox while `runc` will handle
+the `trusted` ones.
```
We might want to mention what happens if you do not mark any as trusted or untrusted, or if you have a mix of trusted/untrusted/un-marked. iirc, we have a table explaining that maybe we could insert here.

sameo commented:
@iphutch Could you please review that PR ?

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+that relies on `OCI` compatible runtimes for managing containers instances.
+
+Clear Containers, being an OCI compatible runtime, naturally fits into the `CRI-O`
+architecture and requirements, and is one of the 2 officially supported `CRI-O` runtimes.
```
s/2/two/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+In other words, `kubelet` is a `CRI` client and expects a `CRI` implementation to
+handle the server side of the interface.
+[CRI-O\*](https://github.com/kubernetes-incubator/cri-o) is a `CRI` implementation
+that relies on `OCI` compatible runtimes for managing containers instances.
```
"container instances" reads slightly better I think.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+### Problem statement
+
+The [Kubernetes\*](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
+i.e. a set of containers sharing the same resources and networking, mount and PID
```
The term "pod" has been defined in the paragraph above.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+The [Kubernetes\*](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
+i.e. a set of containers sharing the same resources and networking, mount and PID
+namespaces. A pod specification is a set of containeris with an associated list of
```
typo: "containeris".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+The [Kubernetes\*](https://github.com/kubernetes/kubernetes/) execution unit is a `Pod`,
+i.e. a set of containers sharing the same resources and networking, mount and PID
+namespaces. A pod specification is a set of containeris with an associated list of
+constraints (namespaces, cgroups, hardware resources, security contexts, etc...).
```
I don't think the ellipsis is required as you've already got "etc" (which strictly should be rendered in italic I think).

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+constraints (namespaces, cgroups, hardware resources, security contexts, etc...).
+By default the Clear Containers runtime associates each container with a dedicated
+virtual machine, but that does not work well with pods as letting a set of virtual
+machiness share the same namespaces and resources can be very complex to implement.
```
typo: "machiness".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+By default the Clear Containers runtime associates each container with a dedicated
+virtual machine, but that does not work well with pods as letting a set of virtual
+machiness share the same namespaces and resources can be very complex to implement.
+For example allowing two virtual machines to share the same networking or PID namespace
```
I'm actually wondering if this sentence could be dropped entirely since what it's stating is very similar to the previous one?

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+the OCI compatible runtime.
+
+Before calling its runtime, `CRI-O` will always add a `io.kubernetes.cri-o.ContainerType`
+annotation to the `config.json` `OCI` configuration file it produces from the kubelet `CRI`
```
I wonder if we can drop the "`OCI` configuration file" text as that has been explained in the previous sentence, so it would become just, "to the `config.json` file it produces..."

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+orchestration support was an afterthought with 2.1, and that supporting the single container
+Docker use case was the main focus.
+
+Starting from the 3.0 release, Clear Containers took a different, almost opposite approach.
```
s/almost opposite approach/almost opposite, approach/

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+One interesting evolution of the `CRI-O` support for the Clear Containers runtime is the ability
+to run virtual machine based pods alongside namespace ones. With `CRI-O` and Clear Containers,
+one can introduce the concept of workload trust inside a Kubernetes cluster.
```
I'm not sure if "workload trust" should in fact have double-quotes around it as it's introducing a term, so will leave to @iphutch to check ;)

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+and Clear Containers based Kubernetes cluster handles this use case transparently as long as
+the deployed containers are properly tagged. All `untrusted` containers will be handled by
+Clear Containers and thus run in a hardware virtualized secure sandbox whilst `runc`, for
+example could  handle the `trusted` ones.
```
I think you need a comma after "example".

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+to run virtual machine based pods alongside namespace ones. With `CRI-O` and Clear Containers,
+one can introduce the concept of workload trust inside a Kubernetes cluster.
+
+A cluster operator can now tag (through Kubernetes annotations) container workloads as `trusted`
```
I appreciate you go on to give an example of these terms, but I wonder if you could add a super-brief summary after the terms to reinforce that it's not CC that is "untrusted", but rather that CC is providing the extra isolation for untrusted apps. Something like the following (but I'm not 100% happy with this wording tbh):

> ... container workloads as `trusted` (for workloads known to be safe) or `untrusted` (for potentially malicious workloads that require the highest degree of isolation).

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+selection is done as follows:
+
+|                                          | `io.kubernetes.cri-o.TrustedSandbox` = N/A | `io.kubernetes.cri-o.TrustedSandbox` = `true` | `io.kubernetes.cri-o.TrustedSandbox` = `false` |
+| :---                                     |     :---:                                  |     :---:                                     |     :---:                                             |
```
Nice summary!

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+runtime will always be `runc`. However, when a pod is **not** `Privileged` the runtime
+selection is done as follows:
+
+|                                          | `io.kubernetes.cri-o.TrustedSandbox` = N/A | `io.kubernetes.cri-o.TrustedSandbox` = `true` | `io.kubernetes.cri-o.TrustedSandbox` = `false` |
```
I'd change the "= N/A" to "not set" or "not specified" for maximum clarity.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+|                                          | `io.kubernetes.cri-o.TrustedSandbox` = N/A | `io.kubernetes.cri-o.TrustedSandbox` = `true` | `io.kubernetes.cri-o.TrustedSandbox` = `false` |
+| :---                                     |     :---:                                  |     :---:                                     |     :---:                                             |
+| Default `CRI-O` trust level: `trusted`   | `runc`                                     | `runc`                                        | Clear Containwers |
```
typo: "Containwers" (and another one below).

sameo commented:
@jodh-intel Many thanks for the thorough review.

jodh-intel commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+runtime will always be `runc`. However, when a pod is **not** `Privileged` the runtime
+selection is done as follows:
+
+|                                          | `io.kubernetes.cri-o.TrustedSandbox` = not set | `io.kubernetes.cri-o.TrustedSandbox` = `true` | `io.kubernetes.cri-o.TrustedSandbox` = `false` |
```
miniscule nit: I'd remove the "=" so that column just shows, "`io.kubernetes.cri-o.TrustedSandbox` not set".

jodh-intel commented:
lgtm.

Just needs a review from @iphutch I think.

iphutch commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+that relies on `OCI` compatible runtimes for managing container instances.
+
+Clear Containers, being an OCI compatible runtime, naturally fits into the `CRI-O`
+architecture and requirements, and is one of the two officially supported `CRI-O` runtimes.
```
Note that by stating "two" we will have to maintain this document and update the number.
To avoid this:

Clear Containers is an officially supported CRI-O runtime. It is OCI compatible and therefore aligns with CRI-Oarchitecture and requirements.

iphutch commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+In other words, `kubelet` is a `CRI` client and expects a `CRI` implementation to
+handle the server side of the interface.
+[CRI-O\*](https://github.com/kubernetes-incubator/cri-o) is a `CRI` implementation
+that relies on `OCI` compatible runtimes for managing container instances.
```
Also, please make OCI a link: \[OCI](https://www.opencontainers.org/about)

iphutch commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+
+### Problem statement
+
+The [Kubernetes\*](https://github.com/kubernetes/kubernetes/) execution unit is a pod,
```
Given we have defined pod, some assumptions can be made. I suggest we replace 640-643 with the following:
The Kubernetes/* execution unit is a pod that has specifications detailing constraints such as namespaces, groups, hardware resources, security contents, etc.

iphutch commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,133 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+i.e. a set of containers sharing the same resources and networking, mount and PID
+namespaces. A pod specification is a set of containers with an associated list of
+constraints (namespaces, cgroups, hardware resources, security contexts, *etc*).
+By default the Clear Containers runtime associates each container with a dedicated
```
We are missing the difference between the environments of the paragraph starting on 644 and the one starting on 650 as both begin with "by default". Is 644 is talking about a node using Docker to start pods? Then is 650 about a node using CRI-O? I suggest we re work both paragraphs.

644 to 648:

By default, Kubernetes uses Docker to start pods and containers within a pod. In these nodes, the Clear Container runtime will associate each container with its own dedicated virtual machine. However, implementing one virtual machine per container is very difficult. The task of providing the Kubernetes pod semantics is complex given the resources of these virtual machines (such as networking or PID) need to be shared.

Then for 650 - 652, I think there is a typo on 151 and "Kubernetes" should be "CRI-O". Is this accurate:

For nodes using the CRI-O runtime, the Clear Containers runtime will still create a virtual machine for each container, however, it can also create and launch one virtual machine per pod. Each container within a pod is then simply a container inside the virtual machine boundary.

sameo commented:
@iphutch I think I addressed all your comments. I cherry picked some of your suggestions for the "problem statement" section and tried to make it more readable/understandable.

jodh-intel commented:
Hi @iphutch - please could you re-review?

iphutch commented on docs/architecture/architecture.md:
```diff
@@ -609,6 +614,140 @@ rootfs by calling `mount(8)` within the counter.  If the devicemapper block devi
+[`gRPC`](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto) 
+based [Container Runtime Interface (CRI)](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node/container-runtime-interface-v1.md).
+
+In other words, a `kubelet` is a `CRI` client and expects a `CRI` implementation to
```
Remove backticks around kubelet, CRI, and OCI. Here, line 637, 638, 640, 674, and 712

sameo replied:
@iphutch backticks removed, thanks.

jodh-intel commented:
Hi @iphutch - I think this is ready for re-review.
